### PR TITLE
docs: add controllers documentation to ARCHITECTURE

### PR DESCRIPTION
**Pull request by howardjohn**

### CODE REVIEW

File: pilot/ARCHITECTURE.md [hunk 1]
```diff
@@ -179,3 +179,162 @@ Despite this expectation in SotW, due to a quirk in the protocol we can actually
...
+
+Istio provides a few helper libraries to get started writing a controller. While these libraries help, there are still a lot of subtleties in correctly writing (and testing!) a controller properly.
+
+To get started writing a controller, review the [Example Controller](pkg/kube/controllers/example_test.go).
+
+### Controllers overview
+
+Below provides a high level overview of controllers in Istiod. For more information about each controller, consulting the controllers Go docs is recommended.
+
+```mermaid
```
Comment (zirain) [inline pilot/ARCHITECTURE.md:199]: does github support this?

**Review by hzxuzhonghu — APPROVED** [review]
Great grapgh, but we may frequently change that, so we may not catch up with code change

File: pilot/ARCHITECTURE.md (hunk 1 above)
Comment (zirain) [reply pilot/ARCHITECTURE.md:199]: preview at https://github.com/howardjohn/istio/blob/arch/controllers/pilot/ARCHITECTURE.md

File: pilot/ARCHITECTURE.md [hunk 2]
```diff
@@ -179,3 +179,162 @@ Despite this expectation in SotW, due to a quirk in the protocol we can actually
 However, all other types *cannot* be deleted explicitly, and instead are cleaned up when all references are removed. This means we can send partial updates for non-root types, without deleting unsent resources. This effectively allows doing delta updates over SotW. This optimization is critical for our endpoints generator, ensuring that when a pod scales we only need to update the endpoints within that pod.
 
 Istio currently supports both SotW and Delta protocol. However, the delta implementation is not yet optimized well, so it performs mostly the same as SotW.
+
+## Controllers
+
+Istiod consists of a collection of controllers. Per Kubernetes, "controllers are control loops that watch the state of your cluster, then make or request changes where needed."
+
+In Istio, we use the term a bit more liberally. Istio controllers watch more than just the start of *a* cluster -- many are reading from multiple clusters, or even external sources (files and XDS). Generally Kubernetes controllers are then writing state back to the cluster; Istio does have a few of these controllers, but most of them are centered around driving the [Proxy Configuration](#proxy-configuration).
```
Comment (dddddai) [inline pilot/ARCHITECTURE.md:187]: It should be "the **_state_** of a cluster"?

**Pull request merged**